platform_secrets/service.py

```diff
@@ -1,28 +1,97 @@
 from __future__ import annotations
 
+import hashlib
 import logging
+import math
 import re
 from dataclasses import dataclass, field
-from typing import Optional
+from typing import Any, Optional
 
-from .kube_client import (
-    KubeClient,
+from apolo_kube_client.errors import (
     ResourceBadRequest,
-    ResourceConflict,
+    ResourceExists,
     ResourceInvalid,
     ResourceNotFound,
 )
 
+from .kube_client import KubeApi
+
 logger = logging.getLogger()
 
 SECRET_API_ORG_LABEL = "platform.neuromation.io/secret-api-org-name"
 APPS_SECRET_NAME = "apps-secrets"
 
 NO_ORG = "NO_ORG"
+NO_ORG_NORMALIZED = "no-org"
 
 NAMESPACE_ORG_LABEL = "platform.apolo.us/org"
 NAMESPACE_PROJECT_LABEL = "platform.apolo.us/project"
 
+KUBE_NAME_LENGTH_MAX = 63
+KUBE_NAMESPACE_SEP = "--"
+KUBE_NAMESPACE_PREFIX = "platform"
+KUBE_NAMESPACE_HASH_LENGTH = 24
+
+
+def generate_namespace_name(org_name: str, project_name: str) -> str:
+    """
+    returns a Kubernetes resource name in the format
+    `platform--<org_name>--<project_name>--<hash>`,
+    ensuring that the total length does not exceed `KUBE_NAME_LENGTH_MAX` characters.
+
+    - `platform--` prefix is never truncated
+    - `<hash>` (a sha256 truncated to 24 chars), is also never truncated
+    - if the names are long, we truncate them evenly,
+      so at least some parts of both org and proj names will remain
+    """
+    if org_name == NO_ORG:
+        org_name = NO_ORG_NORMALIZED
+
+    hashable = f"{org_name}{KUBE_NAMESPACE_SEP}{project_name}"
+    name_hash = hashlib.sha256(hashable.encode("utf-8")).hexdigest()[
+        :KUBE_NAMESPACE_HASH_LENGTH
+    ]
+
+    len_reserved = (
+        len(KUBE_NAMESPACE_PREFIX)
+        + (len(KUBE_NAMESPACE_SEP) * 2)
```

Comment:
```suggestion
        + (len(KUBE_NAMESPACE_SEP) * 3)
```
there are three separators